### ISSUE TITLE
Pissa's [path_initial_model_for_weight_conversion] to build LoRA on top of base model

### ISSUE THREAD

**Issue by mtran14** [huggingface/peft#1929]
Hello,

I have a trained LoRA model that is initialized with the PiSSA method, and other than the trainable LoRA parameters (A,B), the method also alternate the original model weights from W to W', so `h=W'x + BAx`
I would like to reconvert the modified model W' back to the standard model W (please correct me if I'm wrong, but it seems possible using `path_initial_model_for_weight_conversion`) and  according to this [link](https://github.com/GraphPKU/PiSSA?tab=readme-ov-file#convert-pissa-to-lora). So the goal is to find A' and B' so that `h =Wx + B'A'x = W'x + BAx`

Here is my understanding of how to use `path_initial_model_for_weight_conversion`
```
model_name = 'openai/whisper-base'
whisper_model = WhisperForConditionalGeneration.from_pretrained('openai/whisper-base')
config = LoraConfig(
    r=r,
    lora_alpha=64,
    target_modules=['q_proj', 'v_proj'],
    lora_dropout=0.05,
    bias="none",
    use_rslora=True,
    init_lora_weights='pissa'
)
peft_model = get_peft_model(whisper_model, config)
peft_model.peft_config["default"].init_lora_weights = True 
peft_model.save_pretrained("pissa_init") # this should save A_0, B_0; base weights here is W'
...
train peft_model
...
peft_model.save_pretrained(
    'pissa-trained', path_initial_model_for_weight_conversion="pissa_init"
) # this should compute A' and B' from current A, B and the saved A_0, B_0; base weights here is W'
...
# new model
model = WhisperForConditionalGeneration.from_pretrained(model_name)
new_peft_model = PeftModel.from_pretrained(model, 'pissa-trained') # this should load A' and B', which is compatible with original weights W
```
However, the base model weights of `new_peft_model` is different from `whisper-base` official weights. Please correct me if I miss anything in my code, or I misunderstand any concepts.

Thank you for your help!

**Comment by BenjaminBossan** [comment]
For me, this seems to work. I followed the [description here](https://github.com/huggingface/peft/tree/main/examples/pissa_finetuning). Below is a script to reproduce. Note how the weights behave exactly as expected. The line `peft_model.peft_config["default"].init_lora_weights = True` seems to be unnecessary.

```python
import torch
from transformers import WhisperForConditionalGeneration
from peft import get_peft_model, LoraConfig, PeftModel

torch.manual_seed(0)
model_name = 'openai/whisper-base'
whisper_model = WhisperForConditionalGeneration.from_pretrained('openai/whisper-base', device_map=0)
whisper_model.model.encoder.layers[0].self_attn.q_proj.weight.sum().item()
# -1.5700368881225586
# base weight before applying PiSSA

 config = LoraConfig(
    r=32,
    lora_alpha=64,
    target_modules=['q_proj', 'v_proj'],
    lora_dropout=0.05,
    bias="none",
    use_rslora=True,
    init_lora_weights='pissa'
)
peft_model = get_peft_model(whisper_model, config)
peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.base_layer.weight.sum().item()
# -1.8062798976898193
# value of base weight changed as expected

peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.lora_A["default"].weight.sum().item()
# 2.069936752319336
# peft weight

peft_model.save_pretrained("/tmp/peft/issue-1929/init-model")

# dummy training
optimizer = torch.optim.SGD(peft_model.parameters(), lr=1e-7)
inputs = {
    "input_features": torch.randn(4, 80, 3000).cuda(),
    "decoder_input_ids": torch.arange(10).cuda(),
}
for _ in range(3):
    optimizer.zero_grad()
    output = peft_model(**inputs)
    loss = output.logits.sum()
    loss.backward()
    optimizer.step()

peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.base_layer.weight.sum().item()
# -1.8062798976898193
# base weight stayed the same, as expected

peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.lora_A["default"].weight.sum().item()
# 1.2832775115966797
# lora changed, as expected

# store predictions
peft_model.eval()
with torch.inference_mode():
    logits_before = peft_model(**inputs).logits

# reload model
peft_model.save_pretrained(
    "/tmp/peft/issue-1929/trained-model",
    path_initial_model_for_weight_conversion="/tmp/peft/issue-1929/init-model"
)
del whisper_model, peft_model

whisper_model = WhisperForConditionalGeneration.from_pretrained('openai/whisper-base', device_map=0)
new_peft_model = PeftModel.from_pretrained(whisper_model, "/tmp/peft/issue-1929/trained-model")

# compare weights
new_peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.base_layer.weight.sum().item()
# -1.5700368881225586
# base weights are the same as original, as expected
new_peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.lora_A["default"].weight.sum().item()
# 3.35317325592041
# lora weights have adjusted to compensate, as expected
new_peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.lora_A["default"].weight
# 64, 512
# double the rank, as expected

# check that predictions did not change
with torch.inference_mode():
    logits_after = new_peft_model(**inputs).logits
torch.allclose(logits_before, logits_after)
# False
# oh no!!!
```

However, as the last line of the script reveals, even though the weights are correctly loaded, the predictions differ. Why is that? It's because there is a bug with PiSSA when combined with `use_rslora=True`, which results in incorrect LoRA scaling. It is in fact 16, even though rslora should put it at 11.31370849898476.

Thankfully, this can be easily fixed:

```python
# at the start
true_scaling = peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.scaling["default"]

...
# after loading
from peft.tuners.lora import LoraLayer

# override scaling value
for module in new_peft_model.modules():
    if isinstance(module, LoraLayer):
        module.scaling["default"] = true_scaling

with torch.inference_mode():
    logits_scaled = new_peft_model(**inputs).logits

torch.allclose(logits_scaled, logits_before)
# True
# yay!
```

I'll work on a PR to fix the bug.

**Comment by mtran14** [comment]
Thank you for your detailed example. I copied the code and ran it on Google Colab with a T4 GPU, but `new_peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.base_layer.weight.sum().item()` returns `-1.8062798976898193` instead of `-1.5700368881225586`. Here is the code I ran on Colab (copy-pasted from your example above, just changing the save path)
```
import torch
from transformers import WhisperForConditionalGeneration
from peft import get_peft_model, LoraConfig, PeftModel
import os

save_path = '/content/saved-ckpts/'
torch.manual_seed(0)

model_name = 'openai/whisper-base'
whisper_model = WhisperForConditionalGeneration.from_pretrained('openai/whisper-base', device_map=0)
print('Base model', whisper_model.model.encoder.layers[0].self_attn.q_proj.weight.sum().item())
# -1.5700368881225586
# base weight before applying PiSSA

config = LoraConfig(
    r=32,
    lora_alpha=64,
    target_modules=['q_proj', 'v_proj'],
    lora_dropout=0.05,
    bias="none",
    use_rslora=True,
    init_lora_weights='pissa'
)
peft_model = get_peft_model(whisper_model, config)
peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.base_layer.weight.sum().item()
# -1.8062798976898193
# value of base weight changed as expected

peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.lora_A["default"].weight.sum().item()
# 2.069936752319336
# peft weight

peft_model.save_pretrained(os.path.join(save_path, "init-model"))

# dummy training
optimizer = torch.optim.SGD(peft_model.parameters(), lr=1e-7)
inputs = {
    "input_features": torch.randn(4, 80, 3000).cuda(),
    "decoder_input_ids": torch.arange(10).cuda(),
}
for _ in range(3):
    optimizer.zero_grad()
    output = peft_model(**inputs)
    loss = output.logits.sum()
    loss.backward()
    optimizer.step()

peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.base_layer.weight.sum().item()
# -1.8062798976898193
# base weight stayed the same, as expected

peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.lora_A["default"].weight.sum().item()
# 1.2832775115966797
# lora changed, as expected

# store predictions
peft_model.eval()
with torch.inference_mode():
    logits_before = peft_model(**inputs).logits

# reload model
peft_model.save_pretrained(
    os.path.join(save_path, "trained-model"),
    path_initial_model_for_weight_conversion=os.path.join(save_path, "init-model")
)
del whisper_model, peft_model

whisper_model = WhisperForConditionalGeneration.from_pretrained('openai/whisper-base', device_map=0)
new_peft_model = PeftModel.from_pretrained(whisper_model, os.path.join(save_path, "trained-model"))

# compare weights
print('Adjusted model', new_peft_model.base_model.model.model.encoder.layers[0].self_attn.q_proj.base_layer.weight.sum().item())
# expected -1.5700368881225586, but got -1.8062798976898193

```

**Comment by mtran14** [comment]
So this turns out to be a Peft version issue. It seems the latest released pip version of Peft (0.11.1) would return `-1.8062798976898193` with the `new_peft_model` but if you clone and install the Peft repo (i.e., `pip install --editable ./`) then I can successfully reproduce the code.

**Comment by BenjaminBossan** [comment]
> So this turns out to be a Peft version issue

Great that you figured out the cause. The fix for the rslora init is in the works in #1930.

**Comment by cho104** [comment]
Hello @mtran14, 

I'm exploring the use of RsLoRA + PISSA for fine-tuning a Vision Language Model. Since this issue seems to be the only applications of this technique, I was hoping you could share any intuitions or potential setbacks I might encounter. Any advice would be greatly appreciated.
